Source/WebCore/page/Navigation.cpp

```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
     }
 }
 
+static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(Document& document)
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    RefPtr deferredPromise = DeferredPromise::create(globalObject);
+    Ref domPromise = DOMPromise::create(globalObject, *JSC::jsCast<JSC::JSPromise*>(deferredPromise->promise()));
+    return { WTFMove(domPromise), deferredPromise.releaseNonNull() };
+}
+
+// https://webidl.spec.whatwg.org/#wait-for-all
+static void waitForAllPromises(Vector<RefPtr<DOMPromise>>& promises, std::function<void()> fulfilledCallback, std::function<void(JSC::JSValue)> rejectionCallback)
```

Comment:
Can we make promises a const Vector& or a Vector&&?  As it stands now, it looks like an out param, which isn't great.

Also, can we use WTF::Function instead of std::function?

Reply:
> Also, can we use WTF::Function instead of std::function?

In this case since the function is called later it doesn't work out. A ref to `WTF::Function` doesn't work as it goes out of scope and `WTF::Function` is not copyable unlike `std::function`.